compiler/rustc_typeck/src/check/method/suggest.rs

```diff
@@ -975,6 +977,76 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         None
     }
 
+    fn suggest_derive(
+        &self,
+        err: &mut DiagnosticBuilder<'_>,
+        unsatisfied_predicates: &Vec<(ty::Predicate<'tcx>, Option<ty::Predicate<'tcx>>)>,
+    ) {
+        let derivables = [
+            sym::Eq,
+            sym::PartialEq,
+            sym::Ord,
+            sym::PartialOrd,
+            sym::Clone,
+            sym::Copy,
+            sym::Hash,
+            sym::Default,
+            sym::debug_trait,
+        ];
+        let mut derives = unsatisfied_predicates
+            .iter()
+            .filter_map(|(pred, _)| {
+                if let ty::PredicateKind::Trait(trait_pred, _) = pred.kind().skip_binder() {
+                    let trait_ref = trait_pred.trait_ref;
+                    let self_ty = trait_ref.self_ty();
+                    if let ty::Adt(adt_def, _) = self_ty.kind() {
+                        if adt_def.did.is_local() {
+                            let diagnostic_items =
+                                self.tcx.diagnostic_items(trait_ref.def_id.krate);
+                            return derivables.iter().find_map(|trait_derivable| {
+                                if let Some(item_def_id) = diagnostic_items.get(trait_derivable) {
+                                    if item_def_id == &trait_ref.def_id
+                                        && !(adt_def.is_enum() && *trait_derivable == sym::Default)
+                                    {
+                                        return Some((
+                                            format!("{}", self_ty),
+                                            self.tcx.def_span(adt_def.did),
+                                            format!("{}", trait_ref.print_only_trait_path()),
+                                        ));
+                                    }
+                                }
+                                None
+                            });
+                        }
+                    }
+                }
+                None
```

Comment:
This has quite a bit of rightwards drift. Could we instead use `let x = if let ... else { return; }` to early return the closure and make the contents look more linear?

Reply:
Like this?